File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/validationfee/ValidationFeeHijiriQuote.kt
package org.hyperledger.iroha.sdk.validationfee

import java.math.BigInteger
import java.nio.charset.StandardCharsets
import org.hyperledger.iroha.sdk.address.AccountAddress
import org.hyperledger.iroha.sdk.address.AccountAddressException
import org.hyperledger.iroha.sdk.client.JsonParser

/** Current native-Norito Hijiri validation-fee quote layout. */
const val VALIDATION_FEE_HIJIRI_QUOTE_VERSION_V1: Int = 1

/** Maximum aggregate transfer count accepted by the V1 quote route. */
const val VALIDATION_FEE_HIJIRI_QUOTE_MAX_QUALIFYING_TRANSFERS_V1: Int = 100_000

/** Maximum canonical V1 request size accepted by Torii. */
const val VALIDATION_FEE_HIJIRI_QUOTE_MAX_REQUEST_BYTES_V1: Int = 4 * 1024

/** Maximum canonical V1 response size accepted by clients. */
const val VALIDATION_FEE_HIJIRI_QUOTE_MAX_RESPONSE_BYTES_V1: Int = 64 * 1024

/** Stable schema marker returned by the native verifier. */
const val VALIDATION_FEE_HIJIRI_QUOTE_SCHEMA_V1: String =
    "iroha.torii.v1.validation_fee.hijiri_quote.response"

/** Honest assurance marker for a live state-evaluated quote. */
const val VALIDATION_FEE_HIJIRI_QUOTE_ASSURANCE_V1: String =
    "EVALUATED_PROJECTION_NOT_INDEPENDENTLY_WITNESS_VERIFIED"

/** Typed input for one exact aggregate Hijiri validation-fee quote. */
class ValidationFeeHijiriQuoteRequestV1(
    accountId: String,
    qualifyingTransferCount: Int,
) {
    /** Frozen request layout version. */
    @JvmField
    val version: Int = VALIDATION_FEE_HIJIRI_QUOTE_VERSION_V1

    /** Canonical universal account whose effective Hijiri risk is priced. */
    @JvmField
    val accountId: String = requireCanonicalQuoteAccountId(accountId, "accountId")

    /** Number of qualifying transfers aggregated before the single Q16 ceiling. */
    @JvmField
    val qualifyingTransferCount: Int = qualifyingTransferCount.also {
        require(it in 1..VALIDATION_FEE_HIJIRI_QUOTE_MAX_QUALIFYING_TRANSFERS_V1) {
            "qualifyingTransferCount must be between 1 and " +
                VALIDATION_FEE_HIJIRI_QUOTE_MAX_QUALIFYING_TRANSFERS_V1
        }
    }

    /** Encode this request with the authoritative native Norito codec. */
    fun toNoritoBytes(): ByteArray = ValidationFeeHijiriQuoteBridge.encodeRequestV1(this)

    override fun equals(other: Any?): Boolean =
        other is ValidationFeeHijiriQuoteRequestV1 &&
            accountId == other.accountId &&
            qualifyingTransferCount == other.qualifyingTransferCount

    override fun hashCode(): Int = 31 * accountId.hashCode() + qualifyingTransferCount
}

/**
 * Native-verified V1 Hijiri validation-fee quote.
 *
 * Torii evaluates both policy and Hijiri state at one committed snapshot. The assurance marker is
 * intentionally explicit: the live projection is authenticated by the account-signed transport,
 * but is not an independently witness-verified state proof. Admission later binds the advertised
 * policy and Hijiri hashes and rejects a stale quote.
 */
class ValidationFeeHijiriQuoteV1 internal constructor(
    @JvmField val schema: String,
    @JvmField val version: Int,
    @JvmField val assurance: String,
    @JvmField val evaluatedStateHeight: String,
    @JvmField val quotedExecutionHeight: String,
    @JvmField val accountId: String,
    @JvmField val activePolicyVersion: String,
    @JvmField val activePolicyHash: String,
    @JvmField val feeAssetDefinitionId: String,
    @JvmField val treasuryAccountId: String,
    @JvmField val feeScale: Int,
    @JvmField val hijiriParametersVersion: Int,
    @JvmField val hijiriParametersRevision: String,
    @JvmField val hijiriParametersDigest: String,
    @JvmField val defaultAccountRiskQ16: Long,
    @JvmField val effectiveAccountRiskQ16: Long,
    @JvmField val accountRiskRevision: String?,
    @JvmField val accountRiskDigest: String?,
    @JvmField val feeMultiplierQ16: Long,
    @JvmField val hijiriFeeQuoteHash: String,
    @JvmField val basePerTransferFeeMinorUnits: String,
    @JvmField val adjustedPerTransferFeeMinorUnits: String,
    @JvmField val qualifyingTransferCount: Int,
    @JvmField val aggregateBaseFeeMinorUnits: String,
    @JvmField val aggregateAdjustedFeeMinorUnits: String,
) {
    override fun equals(other: Any?): Boolean =
        other is ValidationFeeHijiriQuoteV1 &&
            schema == other.schema &&
            version == other.version &&
            assurance == other.assurance &&
            evaluatedStateHeight == other.evaluatedStateHeight &&
            quotedExecutionHeight == other.quotedExecutionHeight &&
            accountId == other.accountId &&
            activePolicyVersion == other.activePolicyVersion &&
            activePolicyHash == other.activePolicyHash &&
            feeAssetDefinitionId == other.feeAssetDefinitionId &&
            treasuryAccountId == other.treasuryAccountId &&
            feeScale == other.feeScale &&
            hijiriParametersVersion == other.hijiriParametersVersion &&
            hijiriParametersRevision == other.hijiriParametersRevision &&
            hijiriParametersDigest == other.hijiriParametersDigest &&
            defaultAccountRiskQ16 == other.defaultAccountRiskQ16 &&
            effectiveAccountRiskQ16 == other.effectiveAccountRiskQ16 &&
            accountRiskRevision == other.accountRiskRevision &&
            accountRiskDigest == other.accountRiskDigest &&
            feeMultiplierQ16 == other.feeMultiplierQ16 &&
            hijiriFeeQuoteHash == other.hijiriFeeQuoteHash &&
            basePerTransferFeeMinorUnits == other.basePerTransferFeeMinorUnits &&
            adjustedPerTransferFeeMinorUnits == other.adjustedPerTransferFeeMinorUnits &&
            qualifyingTransferCount == other.qualifyingTransferCount &&
            aggregateBaseFeeMinorUnits == other.aggregateBaseFeeMinorUnits &&
            aggregateAdjustedFeeMinorUnits == other.aggregateAdjustedFeeMinorUnits

    override fun hashCode(): Int = listOf(
        schema,
        version,
        assurance,
        evaluatedStateHeight,
        quotedExecutionHeight,
        accountId,
        activePolicyVersion,
        activePolicyHash,
        feeAssetDefinitionId,
        treasuryAccountId,
        feeScale,
        hijiriParametersVersion,
        hijiriParametersRevision,
        hijiriParametersDigest,
        defaultAccountRiskQ16,
        effectiveAccountRiskQ16,
        accountRiskRevision,
        accountRiskDigest,
        feeMultiplierQ16,
        hijiriFeeQuoteHash,
        basePerTransferFeeMinorUnits,
        adjustedPerTransferFeeMinorUnits,
        qualifyingTransferCount,
        aggregateBaseFeeMinorUnits,
        aggregateAdjustedFeeMinorUnits,
    ).hashCode()
}

internal object ValidationFeeHijiriQuoteProjectionParser {
    private val exactFields = setOf(
        "schema",
        "version",
        "assurance",
        "evaluatedStateHeight",
        "quotedExecutionHeight",
        "accountId",
        "activePolicyVersion",
        "activePolicyHash",
        "feeAssetDefinitionId",
        "treasuryAccountId",
        "feeScale",
        "hijiriParametersVersion",
        "hijiriParametersRevision",
        "hijiriParametersDigest",
        "defaultAccountRiskQ16",
        "effectiveAccountRiskQ16",
        "accountRiskRevision",
        "accountRiskDigest",
        "feeMultiplierQ16",
        "hijiriFeeQuoteHash",
        "basePerTransferFeeMinorUnits",
        "adjustedPerTransferFeeMinorUnits",
        "qualifyingTransferCount",
        "aggregateBaseFeeMinorUnits",
        "aggregateAdjustedFeeMinorUnits",
    )

    fun parse(canonicalJsonUtf8: ByteArray): ValidationFeeHijiriQuoteV1 {
        require(canonicalJsonUtf8.isNotEmpty()) {
            "native Hijiri quote verifier returned an empty projection"
        }
        require(canonicalJsonUtf8.size <= VALIDATION_FEE_HIJIRI_QUOTE_MAX_RESPONSE_BYTES_V1) {
            "native Hijiri quote projection exceeds the response bound"
        }
        val text = String(canonicalJsonUtf8, StandardCharsets.UTF_8)
        require(text.toByteArray(StandardCharsets.UTF_8).contentEquals(canonicalJsonUtf8)) {
            "native Hijiri quote projection is not valid UTF-8"
        }
        val parsed = JsonParser.parse(text)
        require(parsed is Map<*, *>) { "native Hijiri quote projection must be an object" }
        @Suppress("UNCHECKED_CAST")
        val root = parsed as Map<String, Any?>
        require(root.keys == exactFields) {
            "native Hijiri quote projection fields differ from the frozen V1 schema"
        }
        val quote = ValidationFeeHijiriQuoteV1(
            schema = requiredString(root, "schema"),
            version = requiredUnsigned(root, "version", 0xffffL).toInt(),
            assurance = requiredString(root, "assurance"),
            evaluatedStateHeight = requiredString(root, "evaluatedStateHeight"),
            quotedExecutionHeight = requiredString(root, "quotedExecutionHeight"),
            accountId = requireCanonicalQuoteAccountId(
                requiredString(root, "accountId"),
                "accountId",
            ),
            activePolicyVersion = requiredString(root, "activePolicyVersion"),
            activePolicyHash = requiredString(root, "activePolicyHash"),
            feeAssetDefinitionId = requiredString(root, "feeAssetDefinitionId"),
            treasuryAccountId = requiredString(root, "treasuryAccountId"),
            feeScale = requiredUnsigned(root, "feeScale", 0xffL).toInt(),
            hijiriParametersVersion =
                requiredUnsigned(root, "hijiriParametersVersion", 0xffffL).toInt(),
            hijiriParametersRevision = requiredString(root, "hijiriParametersRevision"),
            hijiriParametersDigest = requiredString(root, "hijiriParametersDigest"),
            defaultAccountRiskQ16 = requiredUnsigned(root, "defaultAccountRiskQ16", U32_MAX),
            effectiveAccountRiskQ16 = requiredUnsigned(root, "effectiveAccountRiskQ16", U32_MAX),
            accountRiskRevision = optionalString(root, "accountRiskRevision"),
            accountRiskDigest = optionalString(root, "accountRiskDigest"),
            feeMultiplierQ16 = requiredUnsigned(root, "feeMultiplierQ16", U32_MAX),
            hijiriFeeQuoteHash = requiredString(root, "hijiriFeeQuoteHash"),
            basePerTransferFeeMinorUnits = requiredString(root, "basePerTransferFeeMinorUnits"),
            adjustedPerTransferFeeMinorUnits =
                requiredString(root, "adjustedPerTransferFeeMinorUnits"),
            qualifyingTransferCount =
                requiredUnsigned(
                    root,
                    "qualifyingTransferCount",
                    VALIDATION_FEE_HIJIRI_QUOTE_MAX_QUALIFYING_TRANSFERS_V1.toLong(),
                ).toInt(),
            aggregateBaseFeeMinorUnits = requiredString(root, "aggregateBaseFeeMinorUnits"),
            aggregateAdjustedFeeMinorUnits =
                requiredString(root, "aggregateAdjustedFeeMinorUnits"),
        )
        require(quote.schema == VALIDATION_FEE_HIJIRI_QUOTE_SCHEMA_V1) {
            "native Hijiri quote projection has an unsupported schema"
        }
        require(quote.version == VALIDATION_FEE_HIJIRI_QUOTE_VERSION_V1) {
            "native Hijiri quote projection has an unsupported version"
        }
        require(quote.assurance == VALIDATION_FEE_HIJIRI_QUOTE_ASSURANCE_V1) {
            "native Hijiri quote projection has an unsupported assurance marker"
        }
        require((quote.accountRiskRevision == null) == (quote.accountRiskDigest == null)) {
            "native Hijiri quote projection has an incomplete account-risk binding"
        }
        return quote
    }

    private fun requiredString(root: Map<String, Any?>, field: String): String {
        val value = root[field]
        require(value is String && value.isNotEmpty()) {
            "native Hijiri quote projection.$field must be a non-empty string"
        }
        return value
    }

    private fun optionalString(root: Map<String, Any?>, field: String): String? {
        require(root.containsKey(field)) {
            "native Hijiri quote projection.$field is missing"
        }
        val value = root[field]
        require(value == null || value is String && value.isNotEmpty()) {
            "native Hijiri quote projection.$field must be null or a non-empty string"
        }
        return value
    }

    private fun requiredUnsigned(
        root: Map<String, Any?>,
        field: String,
        maximum: Long,
    ): Long {
        val integer = when (val value = root[field]) {
            is Byte -> BigInteger.valueOf(value.toLong())
            is Short -> BigInteger.valueOf(value.toLong())
            is Int -> BigInteger.valueOf(value.toLong())
            is Long -> BigInteger.valueOf(value)
            is BigInteger -> value
            else -> throw IllegalArgumentException(
                "native Hijiri quote projection.$field must be an integer",
            )
        }
        require(integer.signum() >= 0 && integer <= BigInteger.valueOf(maximum)) {
            "native Hijiri quote projection.$field is outside its unsigned range"
        }
        return integer.toLong()
    }

    private const val U32_MAX: Long = 0xffff_ffffL
}

internal fun requireCanonicalQuoteAccountId(value: String, field: String): String {
    require(value.isNotEmpty() && value == value.trim() && value.indexOf('@') < 0) {
        "$field must use one canonical domainless I105 account id"
    }
    val address = try {
        AccountAddress.parseEncodedIgnoringCurveSupport(value, null)
    } catch (error: AccountAddressException) {
        throw IllegalArgumentException(
            "$field must use one canonical domainless I105 account id",
            error,
        )
    }
    val discriminant = AccountAddress.detectI105Discriminant(value)
        ?: throw IllegalArgumentException("$field must use one canonical domainless I105 account id")
    val canonical = try {
        address.toI105(discriminant)
    } catch (error: AccountAddressException) {
        throw IllegalArgumentException(
            "$field must use one canonical domainless I105 account id",
            error,
        )
    }
    require(canonical == value) { "$field must use one canonical domainless I105 account id" }
    return canonical
}

File: IrohaSwift/Sources/IrohaSwift/ValidationFeeHijiriQuote.swift
import Foundation

#if canImport(Darwin)
import Darwin
#endif

/// Errors raised by the authoritative native-Norito Hijiri quote boundary.
public enum ValidationFeeHijiriQuoteError: Error, Equatable, Sendable {
    /// The exact ABI-23 bridge or either additive quote symbol is unavailable.
    case bridgeUnavailable
    /// Native request encoding or response verification rejected the supplied bytes.
    case nativeRejected(Int32)
    /// Native code returned an empty, oversized, or malformed projection.
    case invalidNativeOutput
}

/// Typed input for one aggregate V1 Hijiri validation-fee quote.
public struct ValidationFeeHijiriQuoteRequestV1: Equatable, Sendable {
    /// Current request and response layout.
    public static let version: UInt16 = 1
    /// Largest aggregate transfer count admitted by V1.
    public static let maximumQualifyingTransferCount: UInt32 = 100_000
    /// Largest native-Norito request admitted by the route.
    public static let maximumRequestBytes = 4 * 1_024

    /// Frozen layout version.
    public let version: UInt16
    /// Canonical universal account whose effective Hijiri risk is priced.
    public let accountId: String
    /// Transfers aggregated before the one required Q16 ceiling operation.
    public let qualifyingTransferCount: UInt32

    /// Creates one request bound to an exact canonical I105 account literal.
    public init(accountId: String, qualifyingTransferCount: UInt32) throws {
        guard accountId == accountId.trimmingCharacters(in: .whitespacesAndNewlines),
              !accountId.isEmpty,
              !accountId.contains("@") else {
            throw ToriiClientError.invalidPayload(
                "accountId must be one exact canonical domainless I105 account id"
            )
        }
        do {
            let prefix = try AccountAddress.inspectI105NetworkPrefix(accountId).chainDiscriminant
            let address = try AccountAddress.parseEncoded(accountId, expectedPrefix: prefix)
            guard try address.toI105(networkPrefix: prefix) == accountId else {
                throw ToriiClientError.invalidPayload(
                    "accountId must be one exact canonical domainless I105 account id"
                )
            }
        } catch let error as ToriiClientError {
            throw error
        } catch {
            throw ToriiClientError.invalidPayload(
                "accountId must be one exact canonical domainless I105 account id"
            )
        }
        guard (1...Self.maximumQualifyingTransferCount).contains(qualifyingTransferCount) else {
            throw ToriiClientError.invalidPayload(
                "qualifyingTransferCount must be within 1...100000"
            )
        }
        self.version = Self.version
        self.accountId = accountId
        self.qualifyingTransferCount = qualifyingTransferCount
    }

    /// Encodes this request with the authoritative native Norito implementation.
    public func noritoBytes() throws -> Data {
        try ValidationFeeHijiriQuoteNative.encodeRequestV1(self)
    }
}

/// A native-verified V1 Hijiri validation-fee quote from one committed state snapshot.
///
/// The assurance marker explicitly states that the authenticated live projection is not an
/// independent state witness. Transaction admission later binds the policy and Hijiri hashes and
/// rejects a stale quote.
public struct ValidationFeeHijiriQuoteV1: Decodable, Equatable, Sendable {
    /// Largest native-Norito response accepted by V1 clients.
    public static let maximumResponseBytes = 64 * 1_024
    /// Stable verified projection schema.
    public static let schemaV1 = "iroha.torii.v1.validation_fee.hijiri_quote.response"
    /// Honest assurance marker for a live evaluated projection.
    public static let evaluatedAssuranceV1 =
        "EVALUATED_PROJECTION_NOT_INDEPENDENTLY_WITNESS_VERIFIED"

    public let schema: String
    public let version: UInt16
    public let assurance: String
    public let evaluatedStateHeight: String
    public let quotedExecutionHeight: String
    public let accountId: String
    public let activePolicyVersion: String
    public let activePolicyHash: String
    public let feeAssetDefinitionId: String
    public let treasuryAccountId: String
    public let feeScale: UInt8
    public let hijiriParametersVersion: UInt16
    public let hijiriParametersRevision: String
    public let hijiriParametersDigest: String
    public let defaultAccountRiskQ16: UInt32
    public let effectiveAccountRiskQ16: UInt32
    public let accountRiskRevision: String?
    public let accountRiskDigest: String?
    public let feeMultiplierQ16: UInt32
    public let hijiriFeeQuoteHash: String
    public let basePerTransferFeeMinorUnits: String
    public let adjustedPerTransferFeeMinorUnits: String
    public let qualifyingTransferCount: UInt32
    public let aggregateBaseFeeMinorUnits: String
    public let aggregateAdjustedFeeMinorUnits: String

    private enum CodingKeys: String, CodingKey, CaseIterable {
        case schema
        case version
        case assurance
        case evaluatedStateHeight
        case quotedExecutionHeight
        case accountId
        case activePolicyVersion
        case activePolicyHash
        case feeAssetDefinitionId
        case treasuryAccountId
        case feeScale
        case hijiriParametersVersion
        case hijiriParametersRevision
        case hijiriParametersDigest
        case defaultAccountRiskQ16
        case effectiveAccountRiskQ16
        case accountRiskRevision
        case accountRiskDigest
        case feeMultiplierQ16
        case hijiriFeeQuoteHash
        case basePerTransferFeeMinorUnits
        case adjustedPerTransferFeeMinorUnits
        case qualifyingTransferCount
        case aggregateBaseFeeMinorUnits
        case aggregateAdjustedFeeMinorUnits
    }

    static func parseNativeVerifiedProjection(_ data: Data) throws -> Self {
        guard !data.isEmpty, data.count <= Self.maximumResponseBytes else {
            throw ValidationFeeHijiriQuoteError.invalidNativeOutput
        }
        let object: Any
        do {
            object = try JSONSerialization.jsonObject(with: data)
        } catch {
            throw ToriiClientError.decoding(error)
        }
        guard let dictionary = object as? [String: Any],
              Set(dictionary.keys) == Set(CodingKeys.allCases.map(\.rawValue)) else {
            throw ToriiClientError.invalidPayload(
                "native Hijiri quote projection differs from the frozen V1 field set"
            )
        }
        let quote: Self
        do {
            quote = try JSONDecoder().decode(Self.self, from: data)
        } catch {
            throw ToriiClientError.decoding(error)
        }
        guard quote.schema == Self.schemaV1,
              quote.version == ValidationFeeHijiriQuoteRequestV1.version,
              quote.assurance == Self.evaluatedAssuranceV1,
              (quote.accountRiskRevision == nil) == (quote.accountRiskDigest == nil) else {
            throw ToriiClientError.invalidPayload(
                "native Hijiri quote projection violates the frozen V1 shape"
            )
        }
        return quote
    }
}

/// Authoritative native-Norito codec and verifier for V1 Hijiri quotes.
public enum ValidationFeeHijiriQuoteNative {
    /// Native bridge ABI carrying the additive quote symbols.
    public static let requiredBridgeAbiVersion: UInt32 = 23

    /// Encodes one exact canonical bare-Norito request.
    public static func encodeRequestV1(
        _ request: ValidationFeeHijiriQuoteRequestV1
    ) throws -> Data {
        try NoritoNativeBridge.shared.validationFeeHijiriQuoteRequestV1(request)
    }

    /// Verifies a canonical response against the exact request bytes sent to Torii.
    public static func verifyResponseV1(
        _ responseNorito: Data,
        requestNorito: Data
    ) throws -> ValidationFeeHijiriQuoteV1 {
        let projection = try NoritoNativeBridge.shared
            .validationFeeHijiriQuoteResponseVerifyV1(
                responseNorito,
                requestNorito: requestNorito
            )
        return try ValidationFeeHijiriQuoteV1.parseNativeVerifiedProjection(projection)
    }
}

protocol ValidationFeeHijiriQuoteCoding {
    func encode(_ request: ValidationFeeHijiriQuoteRequestV1) throws -> Data
    func verify(
        _ responseNorito: Data,
        requestNorito: Data
    ) throws -> ValidationFeeHijiriQuoteV1
}

struct NativeValidationFeeHijiriQuoteCodec: ValidationFeeHijiriQuoteCoding {
    func encode(_ request: ValidationFeeHijiriQuoteRequestV1) throws -> Data {
        try ValidationFeeHijiriQuoteNative.encodeRequestV1(request)
    }

    func verify(
        _ responseNorito: Data,
        requestNorito: Data
    ) throws -> ValidationFeeHijiriQuoteV1 {
        try ValidationFeeHijiriQuoteNative.verifyResponseV1(
            responseNorito,
            requestNorito: requestNorito
        )
    }
}

extension NoritoNativeBridge {
    #if canImport(Darwin)
    private typealias ValidationFeeHijiriQuoteRequestFn = @convention(c) (
        UnsafePointer<UInt8>?, CUnsignedLong, UInt32,
        UnsafeMutablePointer<UnsafeMutablePointer<UInt8>?>?,
        UnsafeMutablePointer<CUnsignedLong>?
    ) -> Int32
    private typealias ValidationFeeHijiriQuoteResponseVerifyFn = @convention(c) (
        UnsafePointer<UInt8>?, CUnsignedLong,
        UnsafePointer<UInt8>?, CUnsignedLong,
        UnsafeMutablePointer<UnsafeMutablePointer<UInt8>?>?,
        UnsafeMutablePointer<CUnsignedLong>?
    ) -> Int32
    private typealias ValidationFeeHijiriQuoteFreeFn = @convention(c) (
        UnsafeMutablePointer<UInt8>?
    ) -> Void
    #endif

    func validationFeeHijiriQuoteRequestV1(
        _ request: ValidationFeeHijiriQuoteRequestV1
    ) throws -> Data {
        #if canImport(Darwin)
        guard let function = resolveNativeSymbol(
            "connect_norito_validation_fee_hijiri_quote_request_v1",
            as: ValidationFeeHijiriQuoteRequestFn.self
        ), let free = resolveNativeSymbol(
            "connect_norito_free",
            as: ValidationFeeHijiriQuoteFreeFn.self
        ) else {
            throw ValidationFeeHijiriQuoteError.bridgeUnavailable
        }
        let account = Data(request.accountId.utf8)
        var output: UnsafeMutablePointer<UInt8>?
        var outputLength: CUnsignedLong = 0
        let status = account.withUnsafeBytes { bytes in
            function(
                bytes.bindMemory(to: UInt8.self).baseAddress,
                CUnsignedLong(bytes.count),
                request.qualifyingTransferCount,
                &output,
                &outputLength
            )
        }
        return try Self.copyValidationFeeHijiriQuoteOutput(
            status: status,
            pointer: output,
            length: outputLength,
            maximumBytes: ValidationFeeHijiriQuoteRequestV1.maximumRequestBytes,
            free: free
        )
        #else
        _ = request
        throw ValidationFeeHijiriQuoteError.bridgeUnavailable
        #endif
    }

    func validationFeeHijiriQuoteResponseVerifyV1(
        _ responseNorito: Data,
        requestNorito: Data
    ) throws -> Data {
        guard !responseNorito.isEmpty,
              responseNorito.count <= ValidationFeeHijiriQuoteV1.maximumResponseBytes,
              !requestNorito.isEmpty,
              requestNorito.count <= ValidationFeeHijiriQuoteRequestV1.maximumRequestBytes else {
            throw ToriiClientError.invalidPayload(
                "Hijiri quote request or response exceeds its frozen V1 byte bound"
            )
        }
        #if canImport(Darwin)
        guard let function = resolveNativeSymbol(
            "connect_norito_validation_fee_hijiri_quote_response_verify_v1",
            as: ValidationFeeHijiriQuoteResponseVerifyFn.self
        ), let free = resolveNativeSymbol(
            "connect_norito_free",
            as: ValidationFeeHijiriQuoteFreeFn.self
        ) else {
            throw ValidationFeeHijiriQuoteError.bridgeUnavailable
        }
        var output: UnsafeMutablePointer<UInt8>?
        var outputLength: CUnsignedLong = 0
        let status = responseNorito.withUnsafeBytes { response in
            requestNorito.withUnsafeBytes { request in
                function(
                    response.bindMemory(to: UInt8.self).baseAddress,
                    CUnsignedLong(response.count),
                    request.bindMemory(to: UInt8.self).baseAddress,
                    CUnsignedLong(request.count),
                    &output,
                    &outputLength
                )
            }
        }
        return try Self.copyValidationFeeHijiriQuoteOutput(
            status: status,
            pointer: output,
            length: outputLength,
            maximumBytes: ValidationFeeHijiriQuoteV1.maximumResponseBytes,
            free: free
        )
        #else
        throw ValidationFeeHijiriQuoteError.bridgeUnavailable
        #endif
    }

    #if canImport(Darwin)
    private static func copyValidationFeeHijiriQuoteOutput(
        status: Int32,
        pointer: UnsafeMutablePointer<UInt8>?,
        length: CUnsignedLong,
        maximumBytes: Int,
        free: ValidationFeeHijiriQuoteFreeFn
    ) throws -> Data {
        guard status == 0 else {
            if let pointer { free(pointer) }
            throw ValidationFeeHijiriQuoteError.nativeRejected(status)
        }
        guard let pointer,
              length > 0,
              length <= CUnsignedLong(maximumBytes),
              UInt64(length) <= UInt64(Int.max) else {
            if let pointer { free(pointer) }
            throw ValidationFeeHijiriQuoteError.invalidNativeOutput
        }
        defer { free(pointer) }
        return Data(bytes: pointer, count: Int(length))
    }
    #endif
}
